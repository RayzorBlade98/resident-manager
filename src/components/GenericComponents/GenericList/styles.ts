import { css, Rule } from 'glamor';
import { defaultColors } from '../../../types/styles';

const listContainerStyle: Rule = {
  backgroundColor: defaultColors.mainLight2,
  overflowY: 'auto',
};

const listElementStyle: Rule = {
  width: '100%',
  height: '10%',
  color: defaultColors.mainDark2,
  display: 'flex',
  justifyContent: 'center',
  alignItems: 'center',
  ':hover': {
    backgroundColor: defaultColors.mainLight3,
    cursor: 'pointer',
  },
};

const selectedListElementStyle: Rule = {
  backgroundColor: defaultColors.mainLight3,
};

const styles = {
  listContainer: css(listContainerStyle),
  listElement: css(listElementStyle),
  selectedListElement: css(listElementStyle, selectedListElementStyle),
};

export default styles;
