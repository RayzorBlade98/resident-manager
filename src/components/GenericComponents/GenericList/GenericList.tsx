import { StyleAttribute, merge } from 'glamor';
import React, { PropsWithChildren, Children } from 'react';
import styles from './styles';

interface GenericListProps {
  /**
   * Additional stlye of the list container
   */
  style?: StyleAttribute;

  /**
   * Additional classes of the list container
   */
  className?: string;
}

/**
 * Component that display a list of clickable elements
 */
function GenericList(props: PropsWithChildren<GenericListProps>): JSX.Element {
  const childrenArray = Children.toArray(props.children);
  return (
    <div
      {...merge(styles.listContainer, props.style)}
      /* istanbul ignore next */
      className={`genericList${props.className ? ` ${props.className}` : ''}`}
    >
      {Children.map(childrenArray, (child, index) => (
        <>
          {child}
          {index !== childrenArray.length - 1 && <hr className="hr" />}
        </>
      ))}
    </div>
  );
}

export default GenericList;
