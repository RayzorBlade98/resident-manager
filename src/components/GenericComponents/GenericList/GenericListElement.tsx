import { StyleAttribute, merge } from 'glamor';
import React, { PropsWithChildren } from 'react';
import styles from './styles';

interface GenericListElementProps {
  /**
   * onClick callback
   */
  onClick: () => void;

  /**
   * Whether the element is selected or not
   */
  selected?: boolean;

  /**
   * Additional stlye of the list element
   */
  style?: StyleAttribute;
}

/**
 * Component that displays an element inside of a `GenericList`
 */
function GenericListElement(
  props: PropsWithChildren<GenericListElementProps>,
): JSX.Element {
  return (
    <div
      onClick={props.onClick}
      {...merge(
        props.selected ? styles.selectedListElement : styles.listElement,
        props.style,
      )}
      className={`listElement${props.selected ? ' selectedListElement' : ''}`}
    >
      {props.children}
    </div>
  );
}

export default GenericListElement;
