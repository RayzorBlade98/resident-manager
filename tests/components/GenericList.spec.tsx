/* eslint-disable max-len */

import { RenderResult, render } from '@testing-library/react';
import { css } from 'glamor';
import React from 'react';
import GenericList from '_/components/GenericComponents/GenericList/GenericList';

describe('GenericList', () => {
  let renderResult: RenderResult;

  beforeEach(() => {
    renderResult = render(
      <GenericList style={css({ backgroundColor: 'red' })} className="testList">
        <div>Element 1</div>
        <div>Element 2</div>
        <div>Element 3</div>
      </GenericList>,
    );
  });

  test('should render children and seperation lines correctly', () => {
    // Assert
    const children = renderResult.container.firstElementChild!.children;
    expect(children).toHaveLength(5);
    for (let index = 0; index < children.length; index += 1) {
      const child = children.item(index)!;
      const expectedTag = index % 2 === 0 ? 'DIV' : 'HR';
      expect(child.tagName).toBe(expectedTag);
    }
  });

  test('should match snapshot', () => {
    // Assert
    expect(renderResult.container).toMatchSnapshot();
  });
});
