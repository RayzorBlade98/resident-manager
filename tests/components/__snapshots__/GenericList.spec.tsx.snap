// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenericList should match snapshot 1`] = `
.css-0,
[data-css-0] {
  background-color: red;
  overflow-y: auto;
}

<div>
  <div
    class="genericList testList"
    data-css-0=""
  >
    <div>
      Element 1
    </div>
    <hr
      class="hr"
    />
    <div>
      Element 2
    </div>
    <hr
      class="hr"
    />
    <div>
      Element 3
    </div>
  </div>
</div>
`;
